manager_dashboard/manager_dashboard/js/usergroup-management.js

```diff
@@ -0,0 +1,115 @@
+    nameInput.attr('disabled', true);
+    descriptionInput.attr('disabled', true);
+
+    firebase.database().ref('v2/userGroups/' + name).once("value", function(snapshot) {
+      if (snapshot.exists()) {
+        $('#add-usergroup-modal-loading-message-container').addClass('hidden');
+        nameInput.attr('disabled', false);
+        descriptionInput.attr('disabled', false);
+        alert('A group with this name already exists');
+      } else {
+        var updates = {};
+        updates['/v2/userGroups/' + name] = {
+          userGroupName: name,
+          description,
+        };
+        firebase.database().ref().update(updates, function(error) {
+          $('#add-usergroup-modal-loading-message-container').addClass('hidden');
+          nameInput.attr('disabled', false);
+          descriptionInput.attr('disabled', false);
+
+          if (error) {
+            alert('Failed to add new user');
+            console.info(error);
+          } else {
+            alert('Successfully added new user group');
+            window.location.reload();
+          }
+        });
+      }
+    });
+  } else {
+    nameInput.attr('disabled', false);
+    descriptionInput.attr('disabled', false);
+    $('#add-usergroup-modal-loading-message-container').addClass('hidden');
+    alert('User group name is required and should be less than 50 characters');
+  }
+}
+
+
+function addUserGroupsToTable() {
+  var tableRef = $("#userGroupsTable").DataTable();
+  var rows = [];
+
+  var userGroupsRef = firebase.database().ref("v2/userGroups");
+  userGroupsRef.once('value', function(snapshot) {
+    if (snapshot.exists()) {
+      snapshot.forEach(function(data) {
+        row_array = [];
+        row_array.push(data.val().userGroupName);
+        row_array.push(data.val().description);
+        var usersLength = Object.keys(data.val().users || {}).length;
+
+        if (usersLength > 0) {
+          // add button for user group members
+          btn = document.createElement('button');
+          btn.id = data.key;
+          btn.setAttribute('data-usergroup', JSON.stringify(data.val()));
+          btn.classList.add("btn");
+          btn.classList.add("btn-warning");
+          btn.classList.add("get-user-groups-members");
+          btn.innerHTML = 'getUserGroupMembers()';
+          row_array.push(btn.outerHTML);
+        } else {
+          row_array.push('No members yet');
+        }
+        rows.push(row_array);
+        tableRef.row.add(row_array).draw(false);
+      })
+    }
+  })
+}
+
+function addUsersForSelectedUserGroupToTable() {
+  const userGroup = JSON.parse(this.getAttribute('data-usergroup'));
+  var groupId = this.id;
+  var usersLength = Object.keys(userGroup.users || {}).length;
+
+  var tableRef = $("#groupMembersTable").DataTable();
+  // remove all existing rows
+  tableRef.clear();
+  var rows = []
+
+  var ref = firebase.database().ref("v2/users").orderByChild(`userGroups/${userGroup.userGroupName}`).limitToLast(usersLength);
```

Comment:
This query might return a LOT of data. Each `users` object includes the whole list of their past contributions. For a group with hundreds or thousands of active users, you could be downloading several MB of data here. Is there a way to limit what data is being fetched?

Reply:
Also, there are more than 50k users in production, so you might want to create an index on that child key to keep the query reasonably fast.

Reply:
Sure, I'm figuring out ways to optimize this query

Reply:
Done!